hugr-core/src/extension/const_fold.rs

```diff
@@ -1,19 +1,134 @@
-use std::fmt::Formatter;
-
-use std::fmt::Debug;
+use std::fmt::{Debug, Formatter};
 
 use crate::ops::Value;
-use crate::types::TypeArg;
+use crate::ops::constant::{CustomConst, OpaqueValue, Sum};
+use crate::types::{SumType, TypeArg};
+use crate::{IncomingPort, Node, OutgoingPort, PortIndex};
+
+/// Representation of values used for constant folding.
+/// See [ConstFold], which is used as `dyn` so we cannot parametrize by
+/// [HugrNode](crate::core::HugrNode).
+// Should we be non-exhaustive??
+#[derive(Clone, Debug, PartialEq, Default)]
+pub enum FoldVal {
+    /// Value is unknown, must assume that it could be anything
+    #[default]
+    Unknown,
+    /// A variant of a [SumType]
+    Sum {
+        /// Which variant of the sum type this value is.
+        tag: usize,
+        /// Describes the type of the whole value.
+        // Can we deprecate this immediately? It is only for converting to Value
+        sum_type: SumType,
```

Comment:
Could move these fields to an inner struct and mark them private i.e. `Sum(SumVal)`